### PR TITLE
Application Access Fixes

### LINKED ISSUE #4860: Application Access Issues
`tctl apps` subcommand missing from Enterprise binary.

---

Fix validating application configuration when `teleport.yaml` does not exist.

---

`tctl get cas` does not list JWT signer.

---

Setup a cluster and setup sessions to be recorded at the proxy like the following.

```
auth_service:
  session_recording: "proxy"
  proxy_checks_host_keys: "yes"
```

You see the following error when trying to access an application.

```
ERRO [APP:WEB]   Error forwarding to /, err: user agent getter missing forward/fwd.go:180
ERRO [APP:WEB]   Error forwarding to /favicon.ico, err: user agent getter missing forward/fwd.go:180
```

### PR DESCRIPTION
**Pull request by russjones**
**Description**

Fixes issues found during testing.

**Related Issues**

Fixes https://github.com/gravitational/teleport/issues/4860

### CODE REVIEW

File: lib/reversetunnel/localsite.go [hunk 1]
```diff
@@ -163,13 +163,14 @@ func (s *localSite) DialAuthServer() (conn net.Conn, err error) {
 }
 
 func (s *localSite) Dial(params DialParams) (net.Conn, error) {
-	// If the proxy is in recording mode use the agent to dial and build a
-	// in-memory forwarding server.
+	// If the proxy is in recording mode and a SSH connection is being requested,
+	// use the agent to dial and build a in-memory forwarding server.
```
Comment (webvictim) [inline lib/reversetunnel/localsite.go:167]:
```suggestion
	// use the agent to dial and build an in-memory forwarding server.
```

File: lib/service/cfg.go [hunk 2]
```diff
@@ -541,6 +542,38 @@ type App struct {
...
+	// are invalid subdomains because for trusted clusters the name is used to
+	// construct the domain that the application will be available at.
+	if errs := validation.IsDNS1035Label(a.Name); len(errs) > 0 {
+		return trace.BadParameter("application name %q must be a valid DNS subdomain: https://gravitational.com/teleport/docs/application-access/#application-name", a.Name)
+	}
+	// Parse and validate URL.
+	if _, err := url.Parse(a.URI); err != nil {
+		return trace.Wrap(err)
+	}
+	// If a port was specified or a IP address was provided for the public
```
Comment (webvictim) [inline lib/service/cfg.go:563]:
```suggestion
	// If a port was specified or an IP address was provided for the public
```

File: lib/service/cfg.go [hunk 3]
```diff
@@ -541,6 +542,38 @@ type App struct {
...
+	// address, return an error.
+	if a.PublicAddr != "" {
+		if _, _, err := net.SplitHostPort(a.PublicAddr); err == nil {
+			return trace.BadParameter("public address %q can not contan a port, applications will be available on the same port as the web proxy", a.PublicAddr)
```
Comment (webvictim) [inline lib/service/cfg.go:567]:
```suggestion
			return trace.BadParameter("public address %q can not contain a port, applications will be available on the same port as the web proxy", a.PublicAddr)
```

File: lib/service/cfg_test.go [hunk 4]
```diff
@@ -98,3 +98,48 @@ func (s *ConfigSuite) TestDefaultConfig(c *check.C) {
...
+		inName   string
+		outValid bool
+	}{
+		{
+			desc:     check.Commentf("valid subdomain"),
+			inName:   "foo",
+			outValid: true,
+		},
+		{
+			desc:     check.Commentf("subdomain can not start with a dash"),
```
Comment (webvictim) [inline lib/service/cfg_test.go:115]:
```suggestion
			desc:     check.Commentf("subdomain cannot start with a dash"),
```

File: lib/service/cfg_test.go [hunk 5]
```diff
@@ -98,3 +98,48 @@ func (s *ConfigSuite) TestDefaultConfig(c *check.C) {
...
+			inName:   "-foo",
+			outValid: false,
+		},
+		{
+			desc:     check.Commentf(`subdomain can not contain the exclamation mark character "!"`),
```
Comment (webvictim) [inline lib/service/cfg_test.go:120]:
```suggestion
			desc:     check.Commentf(`subdomain cannot contain the exclamation mark character "!"`),
```

File: lib/config/configuration_test.go [hunk 6]
```diff
@@ -1130,3 +1130,48 @@ app_service:
...
+
+	for _, tt := range tests {
+		clf := CommandLineFlags{
+			AppName: tt.inAppName,
+			AppURI:  tt.inAppURI,
+		}
+		cfg := service.MakeDefaultConfig()
+
+		err := Configure(&clf, cfg)
+		c.Assert(err != nil, check.Equals, tt.outError, tt.desc)
```
Comment (a-palchikov) [inline lib/config/configuration_test.go:1170]: Verify that the error is expected with `check.FitsTypeOf` or `check.ErrorMatches`.

File: lib/reversetunnel/localsite.go (hunk 1 above)
Comment (a-palchikov) [inline lib/reversetunnel/localsite.go:167]: a -> an in-memory

File: lib/reversetunnel/remotesite.go [hunk 7]
```diff
@@ -508,7 +508,8 @@ func (s *remoteSite) Dial(params DialParams) (net.Conn, error) {
 
 	// If the proxy is in recording mode use the agent to dial and build a
```
Comment (a-palchikov) [inline lib/reversetunnel/remotesite.go:509]: Update the comment like above?

File: lib/service/cfg.go [hunk 8]
```diff
@@ -541,6 +542,38 @@ type App struct {
 	Rewrite *Rewrite
 }
 
+// Check validates an application.
+func (a App) Check() error {
+	if a.Name == "" {
+		return trace.BadParameter("missing name")
```
Comment (a-palchikov) [inline lib/service/cfg.go:548]: I'd add context to error messages. If the message is output without context, it is pretty hard to understand where the problem is. Something like `missing application name` can go a long way.

File: lib/service/cfg.go [hunk 9]
```diff
@@ -541,6 +542,38 @@ type App struct {
...
+	}
+	// Check if the application name is a valid subdomain. Don't allow names that
+	// are invalid subdomains because for trusted clusters the name is used to
+	// construct the domain that the application will be available at.
+	if errs := validation.IsDNS1035Label(a.Name); len(errs) > 0 {
+		return trace.BadParameter("application name %q must be a valid DNS subdomain: https://gravitational.com/teleport/docs/application-access/#application-name", a.Name)
+	}
+	// Parse and validate URL.
+	if _, err := url.Parse(a.URI); err != nil {
+		return trace.Wrap(err)
```
Comment (a-palchikov) [inline lib/service/cfg.go:561]: Add error context?

File: lib/service/cfg.go (hunk 3 above)
Comment (a-palchikov) [inline lib/service/cfg.go:567]: conta**i**n
I'd also use the name of the configuration attribute as we do elsewhere to stay consistent (eg. use `public_addr` instead)

File: lib/service/cfg.go [hunk 10]
```diff
@@ -541,6 +542,38 @@ type App struct {
...
+	}
+	// If a port was specified or a IP address was provided for the public
+	// address, return an error.
+	if a.PublicAddr != "" {
+		if _, _, err := net.SplitHostPort(a.PublicAddr); err == nil {
+			return trace.BadParameter("public address %q can not contan a port, applications will be available on the same port as the web proxy", a.PublicAddr)
+		}
+		if net.ParseIP(a.PublicAddr) != nil {
+			return trace.BadParameter("public address %q can not be an IP address, Teleport Application Access uses DNS names for routing", a.PublicAddr)
```
Comment (a-palchikov) [inline lib/service/cfg.go:570]: I think AAP using DNS names for routing is implementation detail that should not matter in the error.

**Review by a-palchikov — APPROVED** [review]

**Review by r0mant — APPROVED** [review]

File: lib/config/configuration.go [hunk 11]
```diff
@@ -1017,29 +1015,33 @@ func Configure(clf *CommandLineFlags, cfg *service.Config) error {
...
-		fileConf.Apps.Service.EnabledFlag = "yes"
+		cfg.Apps.Enabled = true
 
 		// Parse static and dynamic labels.
 		static, dynamic, err := parseLabels(clf.Labels)
 		if err != nil {
 			return trace.BadParameter("labels invalid: %v", err)
 		}
-		dynamicLabels := make([]CommandLabel, 0, len(dynamic))
+		dynamicLabels := make(services.CommandLabels)
```
Comment (r0mant) [inline lib/config/configuration.go:1025]: Actually, one suggestion is I think `parseLabels()` above already returns `services.CommandLabels` so you don't need to reconstruct it again, can just pass `dynamic` below.

**Pull request merged**